### PR TITLE
Fix #2241.

### LINKED ISSUE #2241: Conflict detection for default optionsets
We would like to protect against duplicate optionsets, and need to check for optionsets that already exist in PowerFx as defaults. SymbolTable.TryLookupSlot currently returns false for default optionsets, so we need another way to perform this logic. TIA

### PR DESCRIPTION
**Pull request by MikeStall**
Fix #2241 - adds a TryGetType to lookup a variable by name (very useful when it doesn't have a slot)  and return its type. 

This is particularly interesting for enums/optionsets  - which don't have slots (since they don't have per-eval storage). But 
 beware , #2342  - some of the FormulaTypes for enums/optionsets are not correct.

### CODE REVIEW

File: src/tests/Microsoft.PowerFx.Core.Tests/SymbolTableTests.cs [hunk 1]
```diff
@@ -4,6 +4,7 @@
 using System;
 using System.Collections.Generic;
 using System.Linq;
+using System.Security.Cryptography;
```
Comment (LucGenetier) [inline src/tests/Microsoft.PowerFx.Core.Tests/SymbolTableTests.cs:7]: >System.Security.Cryptography [](http://example.com/codeflow?start=6&length=28)

Why is this needed?

File: src/tests/Microsoft.PowerFx.Core.Tests/SymbolTableTests.cs [hunk 2]
```diff
@@ -310,6 +311,49 @@ public void VoidIsNotAllowed()
...
+                DisplayNameUtility.MakeUnique(new Dictionary<string, string> { { "Yes", "Yes1" }, { "No", "No1" } }));
+
+            PowerFxConfig config = new PowerFxConfig();
+
+            bool fOk = config.SymbolTable.IsDefined("os1");
+            Assert.False(fOk);
+
+            config.AddOptionSet(os);
+
+            fOk = config.SymbolTable.IsDefined("os1");
```
Comment (LucGenetier) [inline src/tests/Microsoft.PowerFx.Core.Tests/SymbolTableTests.cs:330]: >fOk = config.SymbolTable.IsDefined("os1"); [](http://example.com/codeflow?start=12&length=42)

I would add a test with "OS1" to confirm that symbols are case sensitive

**Review by LucGenetier — DISMISSED** [review]
:shipit:

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/tests/Microsoft.PowerFx.Core.Tests/SymbolTableTests.cs (hunk 1 above)
Comment (MikeStall) [reply src/tests/Microsoft.PowerFx.Core.Tests/SymbolTableTests.cs:7]: VS  auto added it - will remove :(

File: src/libraries/Microsoft.PowerFx.Core/Public/Config/ReadOnlySymbolTable.cs [hunk 3]
```diff
@@ -345,6 +345,25 @@ public IEnumerable<NamedFormulaType> SymbolNames
...
         }
 
+        /// <summary>
+        /// Returns true if the symbol is defined. 
+        /// This includes both variables (that would be found with <see cref="TryLookupSlot(string, out ISymbolSlot)"/>, as well as constants (like Enums, OptionSets, constants).
+        /// It does not include Functions since those are in a different namespace. 
+        /// </summary>
+        /// <param name="name">logical name of a symbol.</param>
+        /// <returns></returns>
+        public bool IsDefined(string name)
```
[the remark this answers is not in the document]
Comment (MikeStall) [reply src/libraries/Microsoft.PowerFx.Core/Public/Config/ReadOnlySymbolTable.cs:355]: The problem I'm having is that they want to check things like Color; but that's an Enum and we don't have a public symbol to give back for it.  But we could probably give back FormulaType.

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/RecalcEngineTests.cs [hunk 4]
```diff
@@ -1533,6 +1533,17 @@ public double NextDouble()
...
         }
 
+        [Fact]
+        public void LookupBuiltinOptionSets()
+        {
+            var config = new PowerFxConfig();
+            var engine = new RecalcEngine(config);
+
+            // Builtin enums are on engine.SupportedFunctionm
+            var ok = engine.SupportedFunctions.IsDefined("Color");
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/RecalcEngineTests.cs:1543]: >IsDefined [](http://example.com/codeflow?start=47&length=9)

also give back type

File: src/libraries/Microsoft.PowerFx.Core/Public/Config/ReadOnlySymbolTable.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (MikeStall) [reply src/libraries/Microsoft.PowerFx.Core/Public/Config/ReadOnlySymbolTable.cs:355]: Turns out we hit another bug where FormulaType is wrong for both OptionSets (returns as Records) and Enums (returns as Invalid). 
#2342

So we can give type, but should fix this separately.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Review by LucGenetier — DISMISSED** [review]
:shipit:

**Comment by LucGenetier** [comment]
✅ No public API change.

**Review by Shpakh-MSFT — APPROVED** [review]
:shipit:

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Core/Public/Config/ReadOnlySymbolTable.cs [hunk 5]
```diff
@@ -345,6 +345,28 @@ public IEnumerable<NamedFormulaType> SymbolNames
...
 
+        /// <summary>
+        /// Returns true if the symbol is defined. 
+        /// This includes both variables (that would be found with <see cref="TryLookupSlot(string, out ISymbolSlot)"/>, as well as constants (like Enums, OptionSets, constants).
+        /// It does not include Functions since those are in a different namespace. 
+        /// </summary>
+        /// <param name="name">logical or display name of a symbol.</param>
+        /// <param name="type"></param>
+        /// <returns></returns>
+        public bool TryGetType(string name, out FormulaType type)
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Public/Config/ReadOnlySymbolTable.cs:356]: Due to another bug, this will not give right types for enums for option sets  - see  #2342
But you'd have the same problem if you enumerated the symbols and got the types that way.

**Review by LucGenetier — DISMISSED** [review]
:shipit:

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.
